frontend/lib/src/components/widgets/ButtonGroup/ButtonGroup.tsx

```diff
@@ -0,0 +1,314 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  ReactElement,
+  useEffect,
+  useState,
+  forwardRef,
+  Ref,
+} from "react"
+
+import { useTheme } from "@emotion/react"
+import isEqual from "lodash/isEqual"
+
+import { ButtonGroup as BasewebButtonGroup, MODE } from "baseui/button-group"
+
+import BaseButton, {
+  BaseButtonKind,
+  BaseButtonSize,
+} from "@streamlit/lib/src/components/shared/BaseButton"
+import { DynamicIcon } from "@streamlit/lib/src/components/shared/Icon"
+import { EmotionTheme } from "@streamlit/lib/src/theme"
+
+import { ButtonGroup as ButtonGroupProto } from "@streamlit/lib/src/proto"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import StreamlitMarkdown from "@streamlit/lib/src/components/shared/StreamlitMarkdown"
+import { iconSizes } from "@streamlit/lib/src/theme/primitives"
+import { FormClearHelper } from "@streamlit/lib/src/components/widgets/Form/FormClearHelper"
+
+const materialIconRegexp = /^:material\/(.+):$/
+
+export interface Props {
+  disabled: boolean
+  element: ButtonGroupProto
+  widgetMgr: WidgetStateManager
+  fragmentId?: string
+}
+
+function handleMultiSelection(
+  index: number,
+  currentSelection: number[]
+): number[] {
+  if (!currentSelection.includes(index)) {
+    return [...currentSelection, index]
+  }
+  return currentSelection.filter(value => value !== index)
+}
+
+function handleSelection(
+  mode: ButtonGroupProto.ClickMode,
+  index: number,
+  currentSelection?: number[]
+): number[] {
+  if (mode == ButtonGroupProto.ClickMode.MULTI_SELECT) {
+    return handleMultiSelection(index, currentSelection ?? [])
+  }
+  return [index]
+}
+
+function getSingleSelection(currentSelection: number[]): number {
+  if (currentSelection.length === 0) {
+    return -1
+  }
+  return currentSelection[0]
+}
+
+function syncValue(
+  selected: number[],
+  element: ButtonGroupProto,
+  widgetMgr: WidgetStateManager,
+  fragmentId?: string,
+  fromUi = true
+): void {
+  widgetMgr.setIntArrayValue(element, selected, { fromUi: fromUi }, fragmentId)
+}
+
+function getMaterialIcon(option: string): string | undefined {
```

Comment:
Also: theoretically, you don't even have to apply this extraction since Markdown just supports material icons as part of markdown. So we can just pass all content as is to StreamlitMarkdown.

Reply:
good point, I have renamed it to `isMaterialIcon` and moved it as a shared function to `DynamicIcon`. I didn't know that you [MaterialIcon-for-Markdown support PR](https://github.com/streamlit/streamlit/pull/8889) was merged last week; but I have tried it out and the styling looks weird when rendering just the material icon without any text. It seems to be an issue that a `<p>` tag is used as a wrapper around the material icon's `<span>` tag (vs. `DynamicIcon`, which uses `<span>` tag also as a wrapper. So it looks like the styling would need to be modified, hence I would go with using `DynamicIcon` for now and we can sync about updating the styling next week.